File: array/st/segment_tree.py


# LC3479
class MaxSegmentTree:
	def build(self, node :int, l :int, r :int, arr :List[int]):

		if l == r : self.st[node] = arr[l]
		else:

			m = l + (r -l) //2
			f = node <<1

			self.build(f, l,m, arr)
			self.build(f +1, m +1,r, arr)

			self.st[node] = max(self.st[f],self.st[f+1])

	def update(self, node :int, l :int, r :int, target :int):

		if	l == r:
			if	target <= self.st[node]:
				self.st[node] = 0
				return True
		else:

			m = l + (r -l) //2
			f = node <<1

			if	target <= self.st[f] and self.update(f, l,m, target):

				self.st[node] = max(self.st[f],self.st[f+1])
				return True

			if	target <= self.st[f+1] and self.update(f+1, m +1,r, target):

				self.st[node] = max(self.st[f],self.st[f+1])
				return True

			return False

File: array/st/segmentTree.c


// LC3479
void build(int node, int l, int r, int* st, int* src)
{
	if(l == r) *(st + node) = *(src +l);
	else
	{
		int m = l + (r -l) /2;
		int f = node <<1;

		build(f, l,m, st,src);
		build(f +1, m +1,r, st,src);

		*(st + node) = *(st +f) <*(st +f +1) ? *(st +f +1) : *(st +f);
	}
}
int update(int node, int l, int r, int* st, int T)
{
	if(l == r)
	{
		if(T <= *(st + node))
		{
			*(st + node) = 0;
			return 1;
		}	return 0;
	}
	int m = l + (r -l) /2;
	int f = node <<1;

	if(T <= *(st + f) && update(f, l,m, st, T))
	{
		*(st + node) = *(st +f) <*(st +f +1) ? *(st +f +1) : *(st +f);
		return 1;
	}
	if(T <= *(st +f +1) && update(f +1, m +1,r, st, T))
	{
		*(st + node) = *(st +f) <*(st +f +1) ? *(st +f +1) : *(st +f);
		return 1;
	}	return 0;
}

